### DISCUSSION TITLE
写clib时候有个报错，请问该怎么解决呢

### DISCUSSION THREAD

**Discussion by Scannery** [cloudwu/skynet#1525] [Q&A]
clib引入了skynet.h，直接用g++编译的没问题。
但加上-std=c++11以后会报下面这个错误，应该是malloc声明不匹配。
有什么好点办法可以解决呢，最好是可以不用动两边的源码。

/usr/include/stdlib.h:465:45: 错误：declaration of ‘void* malloc(size_t) throw ()’ has a different exception specifier
 extern void *malloc (size_t __size) __THROW __attribute_malloc__ __wur;
                                             ^
In file included from /home/poa/server/common/skynet/skynet-src/skynet.h:4:0,
                 from /home/poa/server/game/csrc/ser_mymap.cpp:3:
/home/poa/server/common/skynet/skynet-src/skynet_malloc.h:6:23: 错误：从先前的声明‘void* malloc(size_t)’
 #define skynet_malloc malloc

**Comment by cloudwu** [comment]
关闭 malloc hook
https://github.com/cloudwu/skynet/wiki/Build

**Reply by Scannery** [reply]
谢风哥

**Comment by Scannery** [comment]
暂时解决了家人们。
工程里用宏顶掉skynet_malloc.h，自己提供一个套异常壳的实现。
在include skynet.h前引入这个文件。
大概这样：

`#ifndef skynet_malloc_h
#define skynet_malloc_h

#include <stddef.h>
#include <stdlib.h>
extern "C"
{
    //skynet_malloc.h中定义的符号
    void * skynet_malloc(size_t sz);
    void * skynet_calloc(size_t nmemb,size_t size);
    void * skynet_realloc(void *ptr, size_t size);
    void skynet_free(void *ptr);
    char * skynet_strdup(const char *str);
    void * skynet_lalloc(void *ptr, size_t osize, size_t nsize);    // use for lua
    void * skynet_memalign(size_t alignment, size_t size);
    void * skynet_aligned_alloc(size_t alignment, size_t size);
    int skynet_posix_memalign(void **memptr, size_t alignment, size_t size);
}

#define _hook_skynet_malloc malloc
void *_hook_skynet_malloc(size_t __size) __THROW
{
    return skynet_malloc(__size);
}

#define _hook_skynet_calloc calloc
void *_hook_skynet_calloc(size_t __nmemb, size_t __size) __THROW
{
    return skynet_calloc(__nmemb, __size);
}

#define _hook_skynet_realloc realloc
void *_hook_skynet_realloc(void *ptr, size_t size) __THROW
{
    return skynet_realloc(ptr, size);
}

#define _hook_skynet_free free
void _hook_skynet_free(void *ptr) __THROW
{
    skynet_free(ptr);
}

#define _hook_skynet_posix_memalign posix_memalign
int _hook_skynet_posix_memalign(void **memptr, size_t alignment, size_t size) __THROW
{
    return skynet_posix_memalign(memptr, alignment, size);
}

#endif`

**Reply by Scannery** [reply]
nm一下发现应该是hook关了。没有skynet_malloc等符号。
`
#ifndef skynet_malloc_h
#define skynet_malloc_h

#include <stddef.h>
#include <stdlib.h>
#include <iostream>
extern "C"
{
    #define skynet_malloc malloc
    #define skynet_calloc calloc
    #define skynet_realloc realloc 
    #define skynet_free free
    #define skynet_posix_memalign posix_memalign
    #define skynet_memalign memalign
    #define skynet_aligned_alloc aligned_alloc
    
}
`
